{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6de74cdd-e42b-499d-a890-10e770728ab3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import pyfastx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7917efe7-cd14-41bc-b89c-d5c1345ecf2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def telomeres_from_tidehunter(sample: Path):\n",
    "    left = pd.read_csv(list(sample.rglob(\"tidehunter/left_telomeres.csv\"))[0])\n",
    "    right = pd.read_csv(list(sample.rglob(\"tidehunter/right_telomere.csv\"))[0])\n",
    "    both = pd.concat([left, right])\n",
    "    return both\n",
    "\n",
    "\n",
    "def extract_just_telomeres(sample: Path, fraction: float = 0.7):\n",
    "    return (\n",
    "        telomeres_from_tidehunter(sample)\n",
    "        .assign(fraction_telomere=lambda x: x.region_length / x.read_len)\n",
    "        .assign(n_telomeres=lambda x: x.shape[0])\n",
    "        .loc[lambda x: x.fraction_telomere >= fraction]\n",
    "    )\n",
    "\n",
    "def aligned_telomeres_from_tidehunter(sample: Path, NUM_FROM_END = 10_000):\n",
    "    chr_lengths = pd.read_csv(\"chm13v2_chr_lengths.csv\")\n",
    "    telomere_df = (\n",
    "        pd.read_csv(\n",
    "            list(sample.rglob(\"results/*telomere_results.csv\"))[0]\n",
    "        )\n",
    "        .rename(columns={\"region_length\": \"telomere_length\"})\n",
    "        .merge(chr_lengths, on=\"chr\")\n",
    "        .rename(columns={\n",
    "            \"total_length\": \"chr_length\",\n",
    "            }\n",
    "        )\n",
    "        .assign(\n",
    "            arm=lambda x: np.select(\n",
    "                [x.align_start < NUM_FROM_END, x.align_start > x.chr_length - NUM_FROM_END],\n",
    "                [\"left\", \"right\"],\n",
    "                default=pd.NA,\n",
    "            )\n",
    "        )\n",
    "        .assign(fraction_telomere=lambda x: x.telomere_length / x.read_len)\n",
    "        .dropna()\n",
    "        .reset_index()\n",
    "        .sort_values(\"telomere_length\", ascending=False)\n",
    "    )\n",
    "    \n",
    "    return telomere_df\n",
    "    \n",
    "def all_reads_aligned_telomeres(sample: Path, NUM_FROM_END: int = 10_000):\n",
    "    \n",
    "    chr_lengths = pd.read_csv(\"chm13v2_chr_lengths.csv\")\n",
    "    telomere_df = (\n",
    "        pd.read_csv(\n",
    "            list(sample.rglob(\"all_reads_aligned/*all_reads.bed\"))[0],\n",
    "            sep=\"\\t\",\n",
    "            header=None,\n",
    "            names=[\"chr\", \"start\", \"end\", \"id\", \"number\", \"strand\"]\n",
    "        )\n",
    "        .assign(read_length=lambda x: x.end - x.start)\n",
    "        .merge(chr_lengths, on=\"chr\")\n",
    "        .assign(\n",
    "            arm=lambda x: np.select(\n",
    "                [x.start < NUM_FROM_END, x.start > x.total_length - NUM_FROM_END],\n",
    "                [\"left\", \"right\"],\n",
    "                default=pd.NA,\n",
    "            )\n",
    "        )\n",
    "        .dropna()\n",
    "        .reset_index()\n",
    "        .sort_values(\"start\", ascending=True)\n",
    "    )\n",
    "    return telomere_df\n",
    "\n",
    "def attach_fasta_sequence_all(sample: Path):\n",
    "    \n",
    "    telomere_df = all_reads_aligned_telomeres(sample)\n",
    "    fastq_file = list(sample.rglob(\"filtered-fastq/*raw_porechop.fastq.gz\"))[0] \n",
    "    #fastq_file_telomeres = list(sample.rglob(\"telomeres/*telomeres_porechop.fastq.gz\"))[0]\n",
    "    \n",
    "    telomere_ids = telomere_df[\"id\"].values\n",
    "    #telomere_fastq = pyfastx.Fastq(str(fastq_file_telomeres), build_index=False)\n",
    "    telomere_fastq = pyfastx.Fastq(str(fastq_file), build_index=False)\n",
    "    \n",
    "    sequence_name = [read[0] for read in telomere_fastq if read[0] in telomere_ids]\n",
    "    sequence_fasta = [read[1] for read in telomere_fastq if read[0] in telomere_ids]\n",
    "    # PHRED score\n",
    "    #qualities = [list(read[2]) for read in telomere_fastq if read[0] in telomere_ids]\n",
    "    #qualities = [[ord(x) - 33 for x in y] for y in qualities]\n",
    "    \n",
    "    sequence_df =  (\n",
    "        pd.DataFrame()\n",
    "        .assign(\n",
    "            id=sequence_name,\n",
    "            fasta=sequence_fasta,\n",
    "            #PHRED=qualities,\n",
    "        )\n",
    "    )\n",
    "    return sequence_df.merge(telomere_df, on=\"id\")\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "65b94f92-4a35-48e9-b77d-1c325f10727f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "from altair_saver import save\n",
    "def plot_telomeres(sample: Path):\n",
    "    \"\"\"\n",
    "    Plots information about found telomeres\n",
    "    \"\"\"\n",
    "    telomere_df = aligned_telomeres_from_tidehunter(sample)\n",
    "    plot = (\n",
    "        alt.Chart(telomere_df)\n",
    "        .mark_circle()\n",
    "        .encode(\n",
    "            alt.X(\"chr:N\", title=\"Chromosome\"),\n",
    "            alt.Color(\"telomere_type:N\"),\n",
    "            alt.Y(\"telomere_length\", title=\"Length of Telomere\"),\n",
    "            alt.Size(\"read_len:Q\", scale=alt.Scale(range=[50, 500]), title= \"Read length\"),\n",
    "            tooltip=[alt.Tooltip(\"telomere_length:Q\", title=\"Telomere length\"),\n",
    "                     alt.Tooltip(\"telomere_type:N\", title=\"Telomere type\"),\n",
    "                     alt.Tooltip(\"read_len:Q\", title=\"Read length\"),\n",
    "                     alt.Tooltip(\"id:N\", title=\"ID\"),\n",
    "                     alt.Tooltip(\"telomere_pattern_start:Q\", title=\"Start of telomere in read\"),\n",
    "                     alt.Tooltip(\"telomere_pattern_end:Q\", title=\"End of telomere in read\"),\n",
    "                     alt.Tooltip(\"fraction_telomere:Q\", title=\"Fraction of telomere / read\"),\n",
    "                    ]\n",
    "        )\n",
    "        .properties(\n",
    "            height=700,\n",
    "        )\n",
    "        .facet(column=\"arm\", title=f\"Number of telomeric reads: {telomere_df.shape[0]}\")\n",
    "    )\n",
    "    return plot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f6c6cb8-1da5-4580-ba45-93ebc26f62af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to extract fasta file from the above df of telomeres\n",
    "import pyfastx\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "def extract_fasta_telomeres(fastq: str, telomere_df: pd.DataFrame) -> None:\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    telomere_ids = telomere_df[\"id\"].values\n",
    "    telomere_fastq = pyfastx.Fastq(fastq, build_index=False)\n",
    "    \n",
    "    sequence_name = [read[0] for read in telomere_fastq if read[0] in telomere_ids]\n",
    "    sequence_fasta = [read[1] for read in telomere_fastq if read[0] in telomere_ids]\n",
    "    # PHRED score\n",
    "    qualities = [list(read[2]) for read in telomere_fastq if read[0] in telomere_ids]\n",
    "    qualities = [[ord(x) - 33 for x in y] for y in qualities]\n",
    "    \n",
    "    PHRED_df =  (\n",
    "        pd.DataFrame()\n",
    "        .assign(\n",
    "            id=sequence_name,\n",
    "            fasta=sequence_fasta,\n",
    "            PHRED=qualities,\n",
    "        )\n",
    "    )\n",
    "    return PHRED_df.merge(telomere_df, on=\"id\")\n",
    "\n",
    "\n",
    "def plot_telomere_score(phred_df: pd.DataFrame) -> plt.Figure:\n",
    "    \"\"\"\n",
    "    Plots the telomere score\n",
    "    \"\"\"\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_figwidth(15)\n",
    "    ax.plot(phred_df.PHRED)\n",
    "    ax.axvspan(\n",
    "        phred_df.telomere_pattern_start,\n",
    "        phred_df.telomere_pattern_end, \n",
    "        facecolor='lightgreen', \n",
    "        alpha=0.4,\n",
    "    )\n",
    "    ax.set_ylabel('PHRED score')\n",
    "    ax.set_xlabel('Position')\n",
    "    # To put the sequence on the x-axis\n",
    "    #plt.xticks(list(range(len(phred_df.PHRED))), phred_df.fasta, fontsize=5)\n",
    "    \n",
    "    plt.title(f\"\"\"\n",
    "    Read: {phred_df.id}\n",
    "    Motif: {phred_df.telomere_type}\n",
    "    Telomere Length: {phred_df.telomere_length}\n",
    "    \"\"\"\n",
    "    )\n",
    "    return fig, phred_df.fasta\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "#plot.savefig(\"telomere-3'-example.jpg\")\n",
    "\n",
    "# ONTs way to calculate PHRED:\n",
    "# https://labs.epi2me.io/notebooks/Introduction_to_fastq_file.html\n",
    "\n",
    "\n",
    "# open the file and iterate through its records\n",
    "#with FastxFile(\"all_records.fastq\") as fq:\n",
    "#    for rec in fq:\n",
    "#        # ONT calculation for \"mean Q score\"\n",
    "#        quals = np.fromiter(\n",
    "#            (ord(x) - 33 for x in rec.quality),\n",
    "#            dtype=int, count=len(rec.quality))\n",
    "#        mean_p = np.mean(np.power(10, quals/-10))\n",
    "#        mean_qualities.append(-10*np.log10(mean_p))\n",
    "#        # all qualities\n",
    "#        qualities.extend(quals)\n",
    "#        lengths.append(len(quals))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "59f581a8-3e1e-4c2a-8b14-afbddf3412e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_dir = Path(\"/home/nanopore/M622339_ONT_DATA/analyzed-data/\")\n",
    "samples = [x for x in sample_dir.iterdir() if x.is_dir() and not x.stem.startswith(\".\") and not \"Lambda\" in x.stem]\n",
    "\n",
    "for sample in samples[1:]:\n",
    "    \n",
    "    destination = Path(\"/home/nanopore/analyzed-data/dorado-excel-and-plots/\") / sample.stem\n",
    "    destination.mkdir(exist_ok=True)\n",
    "    \n",
    "    # creating files\n",
    "    aligned_telomeres_tidehunter = aligned_telomeres_from_tidehunter(sample)\n",
    "    sequence_df = attach_fasta_sequence_all(sample)\n",
    "    just_telomeres = extract_just_telomeres(sample)\n",
    "    plot = plot_telomeres(sample)\n",
    "    \n",
    "    # saving files\n",
    "    aligned_telomeres_tidehunter.to_excel(f\"{destination}/{sample.stem}_aligned_telomeres_tidehunter.xlsx\")\n",
    "\n",
    "    (\n",
    "        sequence_df\n",
    "        .assign(sequence=lambda x: x.fasta)\n",
    "        .drop(columns=[\"fasta\", \"index\", \"number\", \"total_length\"])\n",
    "        .to_excel(f\"{destination}/{sample.stem}_aligned_to_ends_sequence.xlsx\")\n",
    "\n",
    "    )\n",
    "\n",
    "    just_telomeres.to_excel(f\"{destination}/{sample.stem}_just_telomeres.xlsx\")\n",
    "\n",
    "    save(plot, f\"{destination}/{sample.stem}_tidehunter_aligned_telomere_plot.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45657adb-902a-4690-8e5a-a7d8e03fde73",
   "metadata": {},
   "source": [
    "# New plot and df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "96d7a940-88d5-4358-8450-c9206e7fe019",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_reads_aligned_telomeres_merged_with_tidehunter(sample: Path): \n",
    "\n",
    "    telomeres_tidehunter = (\n",
    "        telomeres_from_tidehunter(sample)\n",
    "        .rename(\n",
    "            columns={\n",
    "                \"start\": \"telomere_start\",\n",
    "                \"end\": \"telomere_end\",\n",
    "                \"region_length\": \"telomere_length\",\n",
    "                \"read_len\": \"read_length\"\n",
    "            }\n",
    "        )\n",
    "        .assign(\n",
    "            telomere_type=lambda x: np.select(\n",
    "                [x.cons_seq.str.contains(\"CCCTAA\"), x.cons_seq.str.contains(\"TTAGGG\")],\n",
    "                [\"CCCTAA\", \"TTAGGG\"]\n",
    "            )\n",
    "        )\n",
    "        .assign(fraction_telomere=lambda x: x.telomere_length / x.read_length)\n",
    "    )\n",
    "    \n",
    "    fasta_df = attach_fasta_sequence_all(sample).drop(columns=\"read_length\")\n",
    "    fasta_df = fasta_df.merge(telomeres_tidehunter, on=\"id\").drop(columns=\"index\")\n",
    "    return fasta_df\n",
    "    \n",
    "\n",
    "def plot_all_reads_aligned_telomeres_merged_with_tidehunter(telomere_df: pd.DataFrame): \n",
    "    plot = (\n",
    "        alt.Chart(telomere_df)\n",
    "        .mark_circle()\n",
    "        .encode(\n",
    "            alt.X(\"chr:N\", title=\"Chromosome\"),\n",
    "            alt.Color(\"telomere_type:N\"),\n",
    "            alt.Y(\"telomere_length\", title=\"Length of Telomere\"),\n",
    "            alt.Size(\"read_length:Q\", scale=alt.Scale(range=[50, 500]), title= \"Read length\"),\n",
    "            tooltip=[alt.Tooltip(\"telomere_length:Q\", title=\"Telomere length\"),\n",
    "                     alt.Tooltip(\"telomere_type:N\", title=\"Telomere type\"),\n",
    "                     alt.Tooltip(\"read_length:Q\", title=\"Read length\"),\n",
    "                     alt.Tooltip(\"id:N\", title=\"ID\"),\n",
    "                     alt.Tooltip(\"telomere_start:Q\", title=\"Start of telomere in read\"),\n",
    "                     alt.Tooltip(\"telomere_end:Q\", title=\"End of telomere in read\"),\n",
    "                     alt.Tooltip(\"fraction_telomere:Q\", title=\"Fraction of telomere / read\"),\n",
    "                    ]\n",
    "        )\n",
    "        .properties(\n",
    "            height=700,\n",
    "        )\n",
    "        .facet(column=\"arm\", title=f\"{sample.stem}, Number telomeres: {telomere_df.shape[0]}\")\n",
    "    )\n",
    "    \n",
    "    return plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "a2c260d1-3b64-4325-be16-644db9a8b7b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_dir = Path(\"/home/nanopore/M622339_ONT_DATA/analyzed-data/\")\n",
    "samples = [x for x in sample_dir.iterdir() if x.is_dir() and not x.stem.startswith(\".\") and not \"Lambda\" in x.stem]\n",
    "\n",
    "for sample in samples:\n",
    "    \n",
    "    destination = Path(\"/home/nanopore/analyzed-data/dorado-excel-and-plots/\") / sample.stem\n",
    "    destination.mkdir(exist_ok=True, parents=True)\n",
    "    \n",
    "    fasta_df = all_reads_aligned_telomeres_merged_with_tidehunter(sample)\n",
    "    plot = plot_all_reads_aligned_telomeres_merged_with_tidehunter(fasta_df) \n",
    "    \n",
    "    fasta_df.to_excel(f\"{destination}/{sample.stem}_aligned_to_ends_merged_tidehunter_and_sequence.xlsx\")\n",
    "    save(plot, f\"{destination}/{sample.stem}_aligned_to_ends_merged_tidehunter_telomere_plot.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dcd4d7d-2704-4215-9d4d-94cf96605dbd",
   "metadata": {},
   "source": [
    "# Telsize tool"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
